constraint/pkg/client/drivers/rego/builtin.go

```diff
@@ -25,12 +31,65 @@ func externalDataBuiltin(d *Driver) func(bctx rego.BuiltinContext, regorequest *
 			return externaldata.HandleError(http.StatusBadRequest, err)
 		}
 
-		externaldataResponse, statusCode, err := d.sendRequestToProvider(bctx.Context, &provider, regoReq.Keys, clientCert)
-		if err != nil {
-			return externaldata.HandleError(statusCode, err)
+		// check provider response cache
+		var providerRequestKeys []string
+		var providerResponseStatusCode int
+		var prepareResponse externaldata.Response
+
+		for _, k := range regoReq.Keys {
+			cachedResponse, err := d.providerResponseCache.Get(
+				externaldata.CacheKey{
+					ProviderName: regoReq.ProviderName,
+					Key:          k,
+				},
+			)
+			if err != nil || time.Since(time.Unix(cachedResponse.Received, 0)) > d.providerResponseCache.TTLInMinutes {
+				// key is not found or cache entry is stale, add key to the provider request keys
+				providerRequestKeys = append(providerRequestKeys, k)
+			} else {
+				prepareResponse.Items = append(
+					prepareResponse.Items, externaldata.Item{
+						Key:   k,
+						Value: cachedResponse.Value,
+						Error: cachedResponse.Error,
+					},
+				)
+			}
+		}
+
+		if len(providerRequestKeys) > 0 {
+			externaldataResponse, statusCode, err := d.sendRequestToProvider(bctx.Context, &provider, providerRequestKeys, clientCert)
+			if err != nil {
+				return externaldata.HandleError(statusCode, err)
+			} else {
+				for _, item := range externaldataResponse.Response.Items {
+					d.providerResponseCache.Upsert(
+						externaldata.CacheKey{
+							ProviderName: regoReq.ProviderName,
+							Key:          item.Key,
+						},
+						externaldata.CacheValue{
+							Received: time.Now().Unix(),
+							Value:    item.Value,
+							Error:    item.Error,
+						},
+					)
+				}
+				prepareResponse.Items = append(prepareResponse.Items, externaldataResponse.Response.Items...)
+			}
+
+			providerResponseStatusCode = statusCode
+			prepareResponse.SystemError = externaldataResponse.Response.SystemError
+		}
+
+		prepareResponse.Idempotent = true
```

Comment:
Why are we setting idempotent to true?

Reply:
Don't we need it to be `true` for mutation?

Reply:
Right, but it's an attestation from the external data source that it is idempotent. Hardcoding that it is defeats the purpose of the check.

Reply:
Got it. Removed.

Reply:
Is the cache being used for mutation as well? We probably want to preserve the response from the provider somehow, if so.

Reply:
@maxsmythe I am now preserving Idempotent from provider response.